### PR TITLE
Remove expr_sub_expressions and rewrite_expression functions

### PR DESCRIPTION
**Pull request by mrob95**
# Which issue does this PR close?

Follows from #2736 and #2765

 # Rationale for this change

These functions were redundant with `ExprVisitor` and `ExprRewriter` and were error prone (as demonstrated in #2736).

# What changes are included in this PR?

Replaced all users of these functions with a visitor or a rewriter as necessary, and deleted them.

Something I'm unsure of: I feel like I'm using `Expr.clone` all over the place - for example in `filter_push_down` at the call sites of `replace_cols_by_name` - but I don't have enough rust experience to know if this is a problem or not.. 

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (andygrove) [inline datafusion/core/src/physical_optimizer/pruning.rs:575]: minor nit: this can be simplified:

```suggestion
            match expr {
                Expr::Column(c) if c == *self.old => Ok(Expr::Column(self.new.clone())),
                _ => Ok(expr)
            }
```

**Review by andygrove — APPROVED** [review]
LGTM. I left one code suggestion. Would be good to get a review from @alamb as well before we merge.

Thanks @mrob95

File: datafusion/optimizer/src/utils.rs
Comment (alamb) [inline datafusion/optimizer/src/utils.rs:32]: 🔥  very nice

**Review by alamb — APPROVED** [review]
This looks great -- what a nice cleanup. Thank you @mrob95 ! 

> Something I'm unsure of: I feel like I'm using Expr.clone all over the place - for example in filter_push_down at the call sites of replace_cols_by_name - but I don't have enough rust experience to know if this is a problem or not..

I think  this is a consequence of the fact that the optimizer passes create new `LogicalPlan`s (rather than rewriting the existing ones). Maybe we can work to improve this over time but it is tricky as the logical plans are sometimes wrapped in `Arc`s (and can be shared).

I don't think this PR adds any new copies, but it makes the cloning more obvious (previously `expr_sub_expression` and `rewrite_expression` internally made all the clones.

**Pull request merged**
